src/charm.py

```diff
@@ -278,6 +278,10 @@ def _update_received_ca_certs(self, event: Optional[CertificateTransferAvailable
         Calling this function from upgrade-charm might be too early though. Pebble-ready is
         preferred.
         """
+        if not self.container.can_connect():
+            event.defer()
```

Comment:
As a general rule of thumb, we try to avoid using defer whenever we can.
https://juju.is/docs/sdk/deferring-events-details-and-dilemmas

I believe this method is also called from pebble-ready, so if the container is not ready when it is first called, it will be called once pebble-ready fires. So we should be good to go without defer?

Reply:
You are right. PR is updated.